my inverter sample data

ivanfmartinez opened the pull request:
I executed the script and saved my json output

Very  different from the other 2 sample files

ivanfmartinez committed:
my inverter sample data

alexdelprete commented:
Hi @ivanfmartinez. Sorry but I just started again thinking about this new approach, and I found out you made this PR a few minutes ago.

Are you still interested to this? What you wrote concerns me a little bit, I gathered data from the VSN700 of a user on HA forum's thread, then I got VSN300 from my device and now I see that there's a third possible dataset.

Can you detail a little your setup? VSN300 card with what fw? Maybe it's due to different versions...

My VSN300 has fw 1.9.2.

ivanfmartinez commented:
Hi @alexdelprete , 

I executed the script to get data and help, the current HA integration using MODBUS is working fine for me. 

But if I have sometime can help here also, now I'm spending some time on my new Luxpower Hybrid inverters.

You can see in the json file that my firmware version is also 1.9.2, AFIK I have the VSN300 model.

ivanfmartinez commented:
For reference I checked that when I started to use the HA integration I have found some differences on my inverter that had to be changed in your original code : https://github.com/alexdelprete/ha-abb-powerone-pvi-sunspec/pull/87

alexdelprete commented:
> You can see in the json file that my firmware version is also 1.9.2, AFIK I have the VSN300 model.

But it's so strange that the output is so different. We have the same card and the same fw version...

ivanfmartinez commented:
> > You can see in the json file that my firmware version is also 1.9.2, AFIK I have the VSN300 model.
> 
> But it's so strange that the output is so different. We have the same card and the same fw version...

probably the output format depends more of the inverter type and than  the interface card.

alexdelprete commented:
I tested it with my VSN300 and inverter (PVI-10.0-OUTD) and a user on the forum (REACT2-5.0-TL with VSN700). There's difference in the data, but the data structures are the same.

So when you run the script the only output is the file you put in the PR? The script calls 3 endpoints /status, /livedata, and /feeds, it then merges the 3 outputs in one file. From what I see from your file, it seems you only get /livedata but I'm not sure.

Could you test the script with the --verbose or --debug switch?

ivanfmartinez commented:
@alexdelprete  As it was a long time I did not remember what I have made before to test and what are the saved data, I updated repository but it failed to get data. 

Put some debug and found that it was trying to use VSN700 basic auth (calling the VSN300 and VSN700 auth handlers but using VSN basic to send request to server).

Changed build_opener use handler based on vsnmodel and was able to get data, don't know if this is the correct aproach for all cases.

```
@@ -74,9 +82,12 @@ class vsnx00Reader():
         self.passman = urllib.request.HTTPPasswordMgrWithPriorAuth()
         self.passman.add_password(self.realm, self.url, self.user, self.password)
         self.logger.debug("Check VSN model: {0}".format(self.vsnmodel))
-        self.handler_vsn300 = VSN300HTTPDigestAuthHandler(self.passman)
-        self.handler_vsn700 = VSN700HTTPPreemptiveBasicAuthHandler(self.passman)
-        self.opener = urllib.request.build_opener(self.handler_vsn700, self.handler_vsn300)
+        if self.vsnmodel == 'vsn300':
+            self.handler_vsn300 = VSN300HTTPDigestAuthHandler(self.passman)
+            self.opener = urllib.request.build_opener(self.handler_vsn300)
+        else:
+            self.handler_vsn700 = VSN700HTTPPreemptiveBasicAuthHandler(self.passman)
+            self.opener = urllib.request.build_opener(self.handler_vsn700)

```

here a full log with debug

[vsnx00-monitor-2025-10-19T102013-0300.log](https://github.com/user-attachments/files/22990454/vsnx00-monitor-2025-10-19T102013-0300.log)

ivanfmartinez commented:
Sorry I just looked that I mixed versions, I will try again with latest from your repo.

ivanfmartinez commented:
@alexdelprete executed with latest version but also have to make some changes, that I don't know how it have to be changed for use on all other inverters.


The handler methods only are called without that context parameter, and with both handlers it does not work.

```
diff --git a/vsnx00-monitor.py b/vsnx00-monitor.py
index 3e44e2b..ec75c5a 100644
--- a/vsnx00-monitor.py
+++ b/vsnx00-monitor.py
@@ -13,11 +13,10 @@ from urllib.request import (HTTPBasicAuthHandler, HTTPDigestAuthHandler,
                             install_opener, parse_http_list, parse_keqv_list,
                             urlopen)
 
-
 def make_request(requested_url):
     try:
         unverified_context = ssl._create_unverified_context()
-        return urlopen(requested_url, timeout=10, context=unverified_context)
+        return urlopen(requested_url, timeout=10)
     except HTTPError as error:
         print(error.status, error.reason)
     except URLError as error:
@@ -62,7 +61,7 @@ class VSN700HTTPPreemptiveBasicAuthHandler(HTTPBasicAuthHandler):
     # Preemptive basic auth: https://stackoverflow.com/a/24048772
     # Instead of waiting for a 403 to then retry with the credentials,
     # send the credentials if the url is handled by the password manager.
-    # Note: please use realm=None when calling add_password
+    # Note: please use realm="" when calling add_password
     def http_request(self, req):
         url = req.get_full_url()
         realm = ""
@@ -101,7 +100,9 @@ class vsnx00Reader():
         self.passman.add_password(self.realm, self.url, self.user, self.password)
         self.handler_vsn300 = VSN300HTTPDigestAuthHandler(self.passman)
         self.handler_vsn700 = VSN700HTTPPreemptiveBasicAuthHandler(self.passman)
-        self.opener = build_opener(self.handler_vsn700, self.handler_vsn300)
+#       need to select the correct handler here or change code to work together
+#        self.opener = build_opener(self.handler_vsn700, self.handler_vsn300)
+        self.opener = build_opener(self.handler_vsn300)
         install_opener(self.opener)

```


[vsnx00-monitor-2025-10-19T110515-0300.log](https://github.com/user-attachments/files/22990655/vsnx00-monitor-2025-10-19T110515-0300.log)

alexdelprete commented:
Strange, when we tried at the time you and Federico reported that it worked.
Have you tried swapping the handlers? Try VSN300 first, then if it fails it tries VSN700 style.

from this:

```python
self.opener = urllib.request.build_opener(self.handler_vsn700, self.handler_vsn300)
```

to this:

```python
self.opener = urllib.request.build_opener(self.handler_vsn300, self.handler_vsn700)
```
